fix HashMap so it no longer hashes all functions using their source stri...

jurko-gospodnetic opened the pull request:
This was causing different function instances created from the same source code
(e.g. as local functions inside the same function) to be hashed the same.

One of the effects was that when using `mocks.module()` and passing it an object
like so:

``` javascript
  mocks.module({myService: 42});
```

worked if it was the first such `mocks.module()` call for a single test. Any later
later such calls (whether passing them same or different object) were simply
ignored by the constructed mocks injector. This is so because all such passed
objects are actually represented as functions in an internal pongular module list
and, since they are all constructed using the same source, they are all hashed
the same so when pongular asks 'do I already have this module loaded?' it thinks
the answer is yes as it can find that same hash in its `HashMap` of already loaded
modules.

Matching workaround has long since been applied in angular sources.

jurko-gospodnetic committed:
fix HashMap so it no longer hashes all functions using their source string

This was causing different function instances created from the same source code
(e.g. as local functions inside the same function) to be hashed the same.

One of the effects was that when using mocks.module() and passing it an object
like so:
  mocks.module({myService: 42});
worked if it was the first such mocks.module() call for a single test. Any later
later such calls (whether passing them same or different object) were simply
ignored by the constructed mocks injector. This is so because all such passed
objects are actually represented as functions in an internal pongular module list
and, since they are all constructed using the same source, they are all hashed
the same so when pongular asks 'do I already have this module loaded?' it thinks
the answer is yes as it can find that same hash in its HashMap of already loaded
modules.

Matching workaround has long since been applied in angular sources.

jurko-gospodnetic commented:
The CI build seems to fail on the old node.js 0.8 but the failure is node.js configuration related and does not have anything to do with the changes made in this pull request.

dvalentiate commented:
Oh this solves my mystery problem! Nice work @jurko-gospodnetic!

I've boiled my code that was breaking to the following. This pull request fixes it! Ya.
```
describe('testing mocks', function () {
	beforeEach(mocks.module({
		theFunctionThatIWant: function () {
			return 'the original result';
		}
	}));
	
	beforeEach(mocks.module({
		theFunctionThatIWant: function () {
			return 'testing result';
		}
	}));
	
	it('should get a function that returns "testing result" when called', mocks.inject(function (theFunctionThatIWant) {
		expect(theFunctionThatIWant()).to.equal('testing result');
	}));
});
```

jurko-gospodnetic committed:
add tests for calling mocks.module() with object notation multiple times

This change also fixes the failing and previously disabled 'should add hashKey to
module function' test inherited from the original angular test suite.

FungusHumungus commented:
Awesome, thanks!

jurko-gospodnetic commented:
Any chance you could push a clean release with this? *nudge* *nudge*

FungusHumungus commented:
0.0.6 published. Thanks for the fix.

On Fri Feb 20 2015 at 9:45:57 AM Stephen Wakely <fungus.humungus@gmail.com>
wrote:

> On the case!
>
>
> On Fri Feb 20 2015 at 9:44:29 AM Jurko Gospodnetić <
> notifications@github.com> wrote:
>
>> Any chance you could push a clean release with this? *nudge* *nudge*
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/FungusHumungus/pongular/pull/20#issuecomment-75212451>
>> .
>>
>

chesleybrown commented on test/pongular-mocksSpec.js:
```diff
@@ -49,19 +49,37 @@ describe('pongular mocks', function() {
 
           }
-          
-          // TODO: test not passing... why?
```
I feel like I remember adding this comment... great to see this fixed! :smile:
